packages/node/src/concordium/api.connection.ts

```diff
@@ -0,0 +1,92 @@
+// Copyright 2020-2023 SubQuery Pte Ltd authors & contributors
+// SPDX-License-Identifier: GPL-3.0
+
+import { EventEmitter2 } from '@nestjs/event-emitter';
+import {
+  ApiConnectionError,
+  ApiErrorType,
+  DisconnectionError,
+  NetworkMetadataPayload,
+  RateLimitError,
+  TimeoutError,
+  IApiConnectionSpecific,
+} from '@subql/node-core';
+import { ConcordiumBlockWrapper } from '@subql/types-concordium';
+import { ConcordiumApi } from './api.concordium';
+import SafeConcordiumGRPCClient from './safe-api';
+
+type FetchFunc = (
+  api: ConcordiumApi,
+  batch: number[],
+) => Promise<ConcordiumBlockWrapper[]>;
+
+export class ConcordiumApiConnection
+  implements
+    IApiConnectionSpecific<
+      ConcordiumApi,
+      SafeConcordiumGRPCClient,
+      ConcordiumBlockWrapper[]
+    >
+{
+  readonly networkMeta: NetworkMetadataPayload;
+
+  constructor(
+    public unsafeApi: ConcordiumApi,
+    private fetchBlocksBatches: FetchFunc,
+  ) {
+    this.networkMeta = {
+      chain: unsafeApi.getChainId().toString(),
+      specName: unsafeApi.getSpecName(),
+      genesisHash: unsafeApi.getGenesisHash(),
+    };
+  }
+
+  static async create(
+    endpoint: string,
+    fetchBlockBatches: FetchFunc,
+    eventEmitter: EventEmitter2,
+  ): Promise<ConcordiumApiConnection> {
+    const api = new ConcordiumApi(endpoint, eventEmitter);
+
+    await api.init();
+
+    return new ConcordiumApiConnection(api, fetchBlockBatches);
+  }
+
+  safeApi(height: number): SafeConcordiumGRPCClient {
+    throw new Error(`Not Implemented`);
```

Comment:
It seems there is an implemented safe api. Does this need to be updated?

Reply:
I think this signature of safeApi should remain unimplemented. we have a seperate method that takes blockHash as param as well.